# Machine Learning Engineer Nanodegree

## Capstone Project - Smile Detector

In this project, CNNs are used to build models to detect if the person in the image is smiling or not.
CelibA dataset is used for this purpose - http://mmlab.ie.cuhk.edu.hk/projects/CelebA.html
The data for this project is downloaded from Kaggle - https://www.kaggle.com/jessicali9530/celeba-dataset

In this notebook, a simple CNN model is developed from scratch. This is considered as Benchmark model based on which improved models will be worked upon.

In [1]:
# all imports
from sklearn.datasets import load_files       
from keras.utils import np_utils
import numpy as np
from glob import glob
from keras.preprocessing import image                  
from tqdm import tqdm

Using TensorFlow backend.


In [5]:
IMG_H=218
IMG_W=178
IMG_D=3

In [6]:
# define function to load train, test, and validation datasets
def load_dataset(path):
    data = load_files(path)
    smile_files = np.array(data['filenames'])
    smile_targets = np_utils.to_categorical(np.array(data['target']), 2)
    return smile_files, smile_targets

# load train, test, and validation datasets
train_files, train_targets = load_dataset('input/dataset/train')
valid_files, valid_targets = load_dataset('input/dataset/validate')
test_files, test_targets = load_dataset('input/dataset/test')

smile_names = [item[:-1] for item in sorted(glob("input/dataset/train/*/"))]

# print statistics about the dataset
print('There are %d total smile categories.' % len(smile_names))
print('There are %s total smile images.\n' % len(np.hstack([train_files, valid_files, test_files])))
print('There are %d training smile images.' % len(train_files))
print('There are %d validation smile images.' % len(valid_files))
print('There are %d test smile images.'% len(test_files))

Using TensorFlow backend.


There are 2 total smile categories.
There are 15000 total smile images.

There are 10000 training smile images.
There are 2500 validation smile images.
There are 2500 test smile images.


In [8]:
def path_to_tensor(img_path):
    # loads RGB image as PIL.Image.Image type
    img = image.load_img(img_path, target_size=(IMG_H, IMG_W))
    # convert PIL.Image.Image type to 3D tensor with shape (IMG_H, IMG_W, IMG_D)
    x = image.img_to_array(img)
    # convert 3D tensor to 4D tensor with shape (1, IMG_H, IMG_W, 3) and return 4D tensor
    return np.expand_dims(x, axis=0)

def paths_to_tensor(img_paths):
    list_of_tensors = [path_to_tensor(img_path) for img_path in tqdm(img_paths)]
    return np.vstack(list_of_tensors)

---

## Creating a CNN to detect smile (from Scratch)

### Pre-process the Data

We rescale the images by dividing every pixel in every image by 255.

In [10]:
from PIL import ImageFile                            
ImageFile.LOAD_TRUNCATED_IMAGES = True                 

# pre-process the data for Keras
train_tensors = paths_to_tensor(train_files).astype('float32')/255
valid_tensors = paths_to_tensor(valid_files).astype('float32')/255
test_tensors = paths_to_tensor(test_files).astype('float32')/255

100%|██████████| 2500/2500 [00:34<00:00, 72.18it/s] 


### BenchModel Architecture
A simple CNN model is developed here considering alternating layers of Conv2D and MaxPooling. Dropout layers are also added to avoid overfitting. All Conv2D layers have relu activation. The final classification dense layer has softmax activation.


In [11]:
from keras.layers import Conv2D, MaxPooling2D, GlobalAveragePooling2D
from keras.layers import Dropout, Flatten, Dense
from keras.models import Sequential

model = Sequential()

model.add(Conv2D(filters=16,kernel_size=2,padding='same',activation='relu',input_shape=(IMG_H,IMG_W,IMG_D)))
model.add(MaxPooling2D(pool_size=2))
model.add(Conv2D(filters=32,kernel_size=2,padding='same',activation='relu'))
model.add(MaxPooling2D(pool_size=2))
model.add(Conv2D(filters=64,kernel_size=2,padding='same',activation='relu'))
model.add(MaxPooling2D(pool_size=2))
model.add(Dropout(0.25))
model.add(Conv2D(filters=128,kernel_size=2,padding='same',activation='relu'))
model.add(MaxPooling2D(pool_size=2))
model.add(Dropout(0.25))
model.add(GlobalAveragePooling2D())
model.add(Dense(2, activation='softmax'))

model.summary()

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_1 (Conv2D)            (None, 218, 178, 16)      208       
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 109, 89, 16)       0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 109, 89, 32)       2080      
_________________________________________________________________
max_pooling2d_3 (MaxPooling2 (None, 54, 44, 32)        0         
_________________________________________________________________
conv2d_3 (Conv2D)            (None, 54, 44, 64)        8256      
_________________________________________________________________
max_pooling2d_4 (MaxPooling2 (None, 27, 22, 64)        0         
_________________________________________________________________
dropout_1 (Dropout)          (None, 27, 22, 64)       

### Compile the Model

In [12]:
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])

### Train the Model

Training the model in the code cell below.  Using model checkpointing to save the model that attains the least validation loss.

In [13]:
from keras.callbacks import ModelCheckpoint  

EPOCHS = 20

checkpointer = ModelCheckpoint(filepath='input/saved_models/weights.best.from_scratch.hdf5', 
                               verbose=1, save_best_only=True)

model.fit(train_tensors, train_targets, 
          validation_data=(valid_tensors, valid_targets),
          epochs=EPOCHS, batch_size=20, callbacks=[checkpointer], verbose=1)

Train on 10000 samples, validate on 2500 samples
Epoch 1/20


 3680/10000 [==========>...................] - ETA: 27:55 - loss: 0.7082 - accuracy: 0.400 - ETA: 17:19 - loss: 0.6722 - accuracy: 0.550 - ETA: 13:15 - loss: 0.6838 - accuracy: 0.550 - ETA: 11:32 - loss: 0.6845 - accuracy: 0.550 - ETA: 12:01 - loss: 0.6835 - accuracy: 0.560 - ETA: 11:04 - loss: 0.6965 - accuracy: 0.525 - ETA: 10:24 - loss: 0.6954 - accuracy: 0.535 - ETA: 9:46 - loss: 0.6935 - accuracy: 0.537 - ETA: 9:18 - loss: 0.7013 - accuracy: 0.50 - ETA: 8:53 - loss: 0.7013 - accuracy: 0.47 - ETA: 8:41 - loss: 0.7000 - accuracy: 0.48 - ETA: 8:42 - loss: 0.6982 - accuracy: 0.50 - ETA: 8:45 - loss: 0.6980 - accuracy: 0.50 - ETA: 8:48 - loss: 0.6972 - accuracy: 0.50 - ETA: 8:35 - loss: 0.6975 - accuracy: 0.50 - ETA: 8:33 - loss: 0.6978 - accuracy: 0.49 - ETA: 8:26 - loss: 0.6971 - accuracy: 0.50 - ETA: 8:18 - loss: 0.6957 - accuracy: 0.51 - ETA: 8:18 - loss: 0.6952 - accuracy: 0.51 - ETA: 8:12 - loss: 0.6932 - accuracy: 0.52 - ETA: 8:07 - loss: 0.6926 - accuracy: 0.52 - ETA: 8:03 - lo

 3720/10000 [==========>...................] - ETA: 5:13 - loss: 0.6981 - accuracy: 0.45 - ETA: 5:12 - loss: 0.6966 - accuracy: 0.47 - ETA: 5:05 - loss: 0.6917 - accuracy: 0.51 - ETA: 4:59 - loss: 0.6929 - accuracy: 0.51 - ETA: 4:55 - loss: 0.6945 - accuracy: 0.50 - ETA: 4:54 - loss: 0.6917 - accuracy: 0.50 - ETA: 5:01 - loss: 0.6923 - accuracy: 0.49 - ETA: 5:09 - loss: 0.6930 - accuracy: 0.48 - ETA: 5:06 - loss: 0.6939 - accuracy: 0.48 - ETA: 5:05 - loss: 0.6929 - accuracy: 0.50 - ETA: 5:04 - loss: 0.6905 - accuracy: 0.50 - ETA: 5:04 - loss: 0.6888 - accuracy: 0.50 - ETA: 5:03 - loss: 0.6855 - accuracy: 0.52 - ETA: 5:33 - loss: 0.6856 - accuracy: 0.52 - ETA: 5:29 - loss: 0.6871 - accuracy: 0.51 - ETA: 5:27 - loss: 0.6871 - accuracy: 0.51 - ETA: 5:24 - loss: 0.6879 - accuracy: 0.51 - ETA: 5:22 - loss: 0.6881 - accuracy: 0.51 - ETA: 5:21 - loss: 0.6894 - accuracy: 0.51 - ETA: 5:20 - loss: 0.6896 - accuracy: 0.51 - ETA: 5:25 - loss: 0.6904 - accuracy: 0.50 - ETA: 5:22 - loss: 0.6906 - ac

10000/10000 [==============================] - ETA: 2:08 - loss: 0.6884 - accuracy: 0.53 - ETA: 2:07 - loss: 0.6893 - accuracy: 0.53 - ETA: 2:06 - loss: 0.6892 - accuracy: 0.53 - ETA: 2:05 - loss: 0.6892 - accuracy: 0.53 - ETA: 2:04 - loss: 0.6892 - accuracy: 0.53 - ETA: 2:03 - loss: 0.6892 - accuracy: 0.53 - ETA: 2:02 - loss: 0.6892 - accuracy: 0.53 - ETA: 2:01 - loss: 0.6892 - accuracy: 0.53 - ETA: 2:00 - loss: 0.6893 - accuracy: 0.53 - ETA: 1:59 - loss: 0.6894 - accuracy: 0.53 - ETA: 1:59 - loss: 0.6894 - accuracy: 0.53 - ETA: 1:58 - loss: 0.6894 - accuracy: 0.53 - ETA: 1:57 - loss: 0.6893 - accuracy: 0.53 - ETA: 1:56 - loss: 0.6893 - accuracy: 0.53 - ETA: 1:55 - loss: 0.6894 - accuracy: 0.53 - ETA: 1:54 - loss: 0.6894 - accuracy: 0.53 - ETA: 1:53 - loss: 0.6894 - accuracy: 0.53 - ETA: 1:52 - loss: 0.6893 - accuracy: 0.53 - ETA: 1:51 - loss: 0.6893 - accuracy: 0.53 - ETA: 1:50 - loss: 0.6893 - accuracy: 0.53 - ETA: 1:49 - loss: 0.6894 - accuracy: 0.53 - ETA: 1:48 - loss: 0.6894 - ac

 3720/10000 [==========>...................] - ETA: 5:00 - loss: 0.7163 - accuracy: 0.50 - ETA: 5:11 - loss: 0.6965 - accuracy: 0.57 - ETA: 5:22 - loss: 0.6845 - accuracy: 0.60 - ETA: 5:22 - loss: 0.6815 - accuracy: 0.60 - ETA: 5:19 - loss: 0.6784 - accuracy: 0.59 - ETA: 5:12 - loss: 0.6841 - accuracy: 0.57 - ETA: 5:06 - loss: 0.6818 - accuracy: 0.58 - ETA: 5:04 - loss: 0.6757 - accuracy: 0.59 - ETA: 5:00 - loss: 0.6832 - accuracy: 0.56 - ETA: 4:58 - loss: 0.6779 - accuracy: 0.57 - ETA: 4:55 - loss: 0.6760 - accuracy: 0.57 - ETA: 4:53 - loss: 0.6728 - accuracy: 0.57 - ETA: 4:51 - loss: 0.6737 - accuracy: 0.57 - ETA: 4:50 - loss: 0.6722 - accuracy: 0.58 - ETA: 4:48 - loss: 0.6714 - accuracy: 0.58 - ETA: 4:47 - loss: 0.6791 - accuracy: 0.57 - ETA: 4:46 - loss: 0.6777 - accuracy: 0.58 - ETA: 4:44 - loss: 0.6761 - accuracy: 0.58 - ETA: 4:44 - loss: 0.6762 - accuracy: 0.58 - ETA: 4:43 - loss: 0.6763 - accuracy: 0.59 - ETA: 4:41 - loss: 0.6747 - accuracy: 0.59 - ETA: 4:40 - loss: 0.6729 - ac

 3720/10000 [==========>...................] - ETA: 4:52 - loss: 0.6763 - accuracy: 0.60 - ETA: 4:49 - loss: 0.6682 - accuracy: 0.60 - ETA: 4:48 - loss: 0.6756 - accuracy: 0.58 - ETA: 4:43 - loss: 0.6800 - accuracy: 0.56 - ETA: 5:33 - loss: 0.6691 - accuracy: 0.60 - ETA: 5:24 - loss: 0.6806 - accuracy: 0.58 - ETA: 5:18 - loss: 0.6787 - accuracy: 0.60 - ETA: 5:12 - loss: 0.6779 - accuracy: 0.58 - ETA: 5:07 - loss: 0.6749 - accuracy: 0.60 - ETA: 5:03 - loss: 0.6712 - accuracy: 0.60 - ETA: 5:00 - loss: 0.6735 - accuracy: 0.60 - ETA: 4:57 - loss: 0.6784 - accuracy: 0.59 - ETA: 4:54 - loss: 0.6781 - accuracy: 0.58 - ETA: 4:52 - loss: 0.6798 - accuracy: 0.57 - ETA: 4:50 - loss: 0.6811 - accuracy: 0.56 - ETA: 4:48 - loss: 0.6832 - accuracy: 0.56 - ETA: 4:47 - loss: 0.6826 - accuracy: 0.55 - ETA: 4:45 - loss: 0.6842 - accuracy: 0.54 - ETA: 4:44 - loss: 0.6827 - accuracy: 0.55 - ETA: 4:43 - loss: 0.6808 - accuracy: 0.55 - ETA: 4:41 - loss: 0.6799 - accuracy: 0.55 - ETA: 4:40 - loss: 0.6795 - ac

10000/10000 [==============================] - ETA: 1:15 - loss: 0.6801 - accuracy: 0.56 - ETA: 1:14 - loss: 0.6801 - accuracy: 0.56 - ETA: 1:13 - loss: 0.6801 - accuracy: 0.56 - ETA: 1:13 - loss: 0.6801 - accuracy: 0.56 - ETA: 1:12 - loss: 0.6800 - accuracy: 0.56 - ETA: 1:12 - loss: 0.6799 - accuracy: 0.56 - ETA: 1:11 - loss: 0.6799 - accuracy: 0.56 - ETA: 1:10 - loss: 0.6800 - accuracy: 0.56 - ETA: 1:10 - loss: 0.6800 - accuracy: 0.56 - ETA: 1:09 - loss: 0.6800 - accuracy: 0.56 - ETA: 1:09 - loss: 0.6799 - accuracy: 0.56 - ETA: 1:08 - loss: 0.6801 - accuracy: 0.56 - ETA: 1:07 - loss: 0.6800 - accuracy: 0.56 - ETA: 1:07 - loss: 0.6802 - accuracy: 0.56 - ETA: 1:06 - loss: 0.6801 - accuracy: 0.56 - ETA: 1:06 - loss: 0.6801 - accuracy: 0.56 - ETA: 1:05 - loss: 0.6800 - accuracy: 0.56 - ETA: 1:05 - loss: 0.6800 - accuracy: 0.56 - ETA: 1:04 - loss: 0.6801 - accuracy: 0.56 - ETA: 1:03 - loss: 0.6800 - accuracy: 0.56 - ETA: 1:03 - loss: 0.6799 - accuracy: 0.56 - ETA: 1:02 - loss: 0.6800 - ac

 3720/10000 [==========>...................] - ETA: 6:41 - loss: 0.6790 - accuracy: 0.60 - ETA: 6:28 - loss: 0.6519 - accuracy: 0.62 - ETA: 6:05 - loss: 0.6788 - accuracy: 0.53 - ETA: 5:42 - loss: 0.6723 - accuracy: 0.56 - ETA: 5:29 - loss: 0.6602 - accuracy: 0.58 - ETA: 5:20 - loss: 0.6555 - accuracy: 0.58 - ETA: 5:14 - loss: 0.6576 - accuracy: 0.56 - ETA: 5:08 - loss: 0.6644 - accuracy: 0.55 - ETA: 5:07 - loss: 0.6689 - accuracy: 0.55 - ETA: 5:04 - loss: 0.6691 - accuracy: 0.55 - ETA: 5:01 - loss: 0.6715 - accuracy: 0.55 - ETA: 4:58 - loss: 0.6715 - accuracy: 0.54 - ETA: 4:56 - loss: 0.6711 - accuracy: 0.55 - ETA: 4:54 - loss: 0.6763 - accuracy: 0.54 - ETA: 4:53 - loss: 0.6759 - accuracy: 0.55 - ETA: 4:52 - loss: 0.6696 - accuracy: 0.56 - ETA: 4:49 - loss: 0.6740 - accuracy: 0.55 - ETA: 4:48 - loss: 0.6768 - accuracy: 0.55 - ETA: 5:07 - loss: 0.6732 - accuracy: 0.57 - ETA: 5:04 - loss: 0.6737 - accuracy: 0.56 - ETA: 5:02 - loss: 0.6763 - accuracy: 0.56 - ETA: 5:03 - loss: 0.6736 - ac

 3720/10000 [==========>...................] - ETA: 5:16 - loss: 0.6858 - accuracy: 0.55 - ETA: 6:06 - loss: 0.6759 - accuracy: 0.62 - ETA: 5:42 - loss: 0.6912 - accuracy: 0.58 - ETA: 5:35 - loss: 0.6820 - accuracy: 0.63 - ETA: 5:21 - loss: 0.6809 - accuracy: 0.63 - ETA: 5:46 - loss: 0.6818 - accuracy: 0.61 - ETA: 5:37 - loss: 0.6812 - accuracy: 0.62 - ETA: 5:48 - loss: 0.6740 - accuracy: 0.63 - ETA: 5:41 - loss: 0.6662 - accuracy: 0.64 - ETA: 5:44 - loss: 0.6531 - accuracy: 0.66 - ETA: 5:55 - loss: 0.6603 - accuracy: 0.64 - ETA: 5:48 - loss: 0.6602 - accuracy: 0.64 - ETA: 5:43 - loss: 0.6565 - accuracy: 0.65 - ETA: 5:42 - loss: 0.6625 - accuracy: 0.63 - ETA: 5:44 - loss: 0.6643 - accuracy: 0.63 - ETA: 5:43 - loss: 0.6599 - accuracy: 0.64 - ETA: 5:55 - loss: 0.6615 - accuracy: 0.63 - ETA: 6:09 - loss: 0.6638 - accuracy: 0.63 - ETA: 6:05 - loss: 0.6596 - accuracy: 0.63 - ETA: 6:08 - loss: 0.6567 - accuracy: 0.63 - ETA: 7:28 - loss: 0.6585 - accuracy: 0.63 - ETA: 7:52 - loss: 0.6556 - ac

10000/10000 [==============================] - ETA: 1:42 - loss: 0.6665 - accuracy: 0.59 - ETA: 1:42 - loss: 0.6664 - accuracy: 0.59 - ETA: 1:41 - loss: 0.6664 - accuracy: 0.59 - ETA: 1:40 - loss: 0.6669 - accuracy: 0.59 - ETA: 1:39 - loss: 0.6667 - accuracy: 0.59 - ETA: 1:38 - loss: 0.6667 - accuracy: 0.59 - ETA: 1:37 - loss: 0.6669 - accuracy: 0.59 - ETA: 1:36 - loss: 0.6668 - accuracy: 0.59 - ETA: 1:36 - loss: 0.6672 - accuracy: 0.59 - ETA: 1:35 - loss: 0.6671 - accuracy: 0.59 - ETA: 1:34 - loss: 0.6672 - accuracy: 0.59 - ETA: 1:33 - loss: 0.6671 - accuracy: 0.59 - ETA: 1:32 - loss: 0.6669 - accuracy: 0.59 - ETA: 1:32 - loss: 0.6668 - accuracy: 0.59 - ETA: 1:31 - loss: 0.6666 - accuracy: 0.59 - ETA: 1:30 - loss: 0.6666 - accuracy: 0.59 - ETA: 1:29 - loss: 0.6666 - accuracy: 0.59 - ETA: 1:28 - loss: 0.6667 - accuracy: 0.59 - ETA: 1:27 - loss: 0.6670 - accuracy: 0.59 - ETA: 1:27 - loss: 0.6671 - accuracy: 0.59 - ETA: 1:26 - loss: 0.6672 - accuracy: 0.59 - ETA: 1:25 - loss: 0.6671 - ac

 3720/10000 [==========>...................] - ETA: 5:07 - loss: 0.6687 - accuracy: 0.65 - ETA: 5:00 - loss: 0.6524 - accuracy: 0.60 - ETA: 4:59 - loss: 0.7034 - accuracy: 0.53 - ETA: 7:18 - loss: 0.6832 - accuracy: 0.56 - ETA: 6:41 - loss: 0.6794 - accuracy: 0.55 - ETA: 6:19 - loss: 0.6895 - accuracy: 0.54 - ETA: 6:08 - loss: 0.7009 - accuracy: 0.55 - ETA: 5:59 - loss: 0.6960 - accuracy: 0.55 - ETA: 5:52 - loss: 0.6914 - accuracy: 0.55 - ETA: 5:52 - loss: 0.6843 - accuracy: 0.56 - ETA: 5:47 - loss: 0.6971 - accuracy: 0.53 - ETA: 5:41 - loss: 0.6932 - accuracy: 0.54 - ETA: 5:36 - loss: 0.6880 - accuracy: 0.55 - ETA: 5:32 - loss: 0.6905 - accuracy: 0.55 - ETA: 5:27 - loss: 0.6906 - accuracy: 0.54 - ETA: 5:24 - loss: 0.6876 - accuracy: 0.54 - ETA: 5:20 - loss: 0.6854 - accuracy: 0.54 - ETA: 5:17 - loss: 0.6796 - accuracy: 0.54 - ETA: 5:15 - loss: 0.6792 - accuracy: 0.54 - ETA: 5:12 - loss: 0.6823 - accuracy: 0.54 - ETA: 5:09 - loss: 0.6835 - accuracy: 0.54 - ETA: 5:09 - loss: 0.6814 - ac

 3720/10000 [==========>...................] - ETA: 5:55 - loss: 0.6834 - accuracy: 0.60 - ETA: 5:21 - loss: 0.6298 - accuracy: 0.67 - ETA: 5:15 - loss: 0.6823 - accuracy: 0.63 - ETA: 5:10 - loss: 0.6774 - accuracy: 0.62 - ETA: 5:07 - loss: 0.6735 - accuracy: 0.62 - ETA: 5:03 - loss: 0.6650 - accuracy: 0.61 - ETA: 4:59 - loss: 0.6557 - accuracy: 0.64 - ETA: 5:16 - loss: 0.6538 - accuracy: 0.65 - ETA: 5:10 - loss: 0.6508 - accuracy: 0.65 - ETA: 5:04 - loss: 0.6541 - accuracy: 0.64 - ETA: 5:01 - loss: 0.6598 - accuracy: 0.64 - ETA: 5:07 - loss: 0.6657 - accuracy: 0.62 - ETA: 5:03 - loss: 0.6696 - accuracy: 0.62 - ETA: 5:02 - loss: 0.6697 - accuracy: 0.61 - ETA: 5:00 - loss: 0.6638 - accuracy: 0.62 - ETA: 4:58 - loss: 0.6605 - accuracy: 0.62 - ETA: 4:56 - loss: 0.6608 - accuracy: 0.62 - ETA: 4:55 - loss: 0.6580 - accuracy: 0.62 - ETA: 4:53 - loss: 0.6571 - accuracy: 0.62 - ETA: 4:51 - loss: 0.6591 - accuracy: 0.61 - ETA: 4:49 - loss: 0.6709 - accuracy: 0.60 - ETA: 4:48 - loss: 0.6711 - ac

10000/10000 [==============================] - ETA: 1:18 - loss: 0.6568 - accuracy: 0.61 - ETA: 1:17 - loss: 0.6566 - accuracy: 0.61 - ETA: 1:16 - loss: 0.6567 - accuracy: 0.61 - ETA: 1:16 - loss: 0.6567 - accuracy: 0.61 - ETA: 1:15 - loss: 0.6566 - accuracy: 0.61 - ETA: 1:15 - loss: 0.6566 - accuracy: 0.61 - ETA: 1:14 - loss: 0.6565 - accuracy: 0.61 - ETA: 1:13 - loss: 0.6569 - accuracy: 0.61 - ETA: 1:13 - loss: 0.6570 - accuracy: 0.61 - ETA: 1:12 - loss: 0.6569 - accuracy: 0.61 - ETA: 1:11 - loss: 0.6571 - accuracy: 0.61 - ETA: 1:11 - loss: 0.6570 - accuracy: 0.61 - ETA: 1:10 - loss: 0.6572 - accuracy: 0.61 - ETA: 1:10 - loss: 0.6578 - accuracy: 0.61 - ETA: 1:09 - loss: 0.6578 - accuracy: 0.61 - ETA: 1:08 - loss: 0.6578 - accuracy: 0.61 - ETA: 1:08 - loss: 0.6576 - accuracy: 0.61 - ETA: 1:07 - loss: 0.6577 - accuracy: 0.61 - ETA: 1:06 - loss: 0.6576 - accuracy: 0.61 - ETA: 1:06 - loss: 0.6574 - accuracy: 0.61 - ETA: 1:05 - loss: 0.6574 - accuracy: 0.61 - ETA: 1:05 - loss: 0.6573 - ac

 3720/10000 [==========>...................] - ETA: 4:42 - loss: 0.5991 - accuracy: 0.65 - ETA: 4:37 - loss: 0.6477 - accuracy: 0.62 - ETA: 4:35 - loss: 0.6082 - accuracy: 0.68 - ETA: 4:32 - loss: 0.6144 - accuracy: 0.67 - ETA: 4:33 - loss: 0.6657 - accuracy: 0.66 - ETA: 4:32 - loss: 0.6540 - accuracy: 0.65 - ETA: 4:31 - loss: 0.6525 - accuracy: 0.64 - ETA: 4:32 - loss: 0.6473 - accuracy: 0.63 - ETA: 4:30 - loss: 0.6412 - accuracy: 0.64 - ETA: 4:30 - loss: 0.6447 - accuracy: 0.63 - ETA: 4:29 - loss: 0.6496 - accuracy: 0.62 - ETA: 4:29 - loss: 0.6420 - accuracy: 0.64 - ETA: 4:28 - loss: 0.6503 - accuracy: 0.63 - ETA: 4:28 - loss: 0.6546 - accuracy: 0.62 - ETA: 4:27 - loss: 0.6546 - accuracy: 0.62 - ETA: 4:27 - loss: 0.6564 - accuracy: 0.62 - ETA: 4:26 - loss: 0.6585 - accuracy: 0.61 - ETA: 4:28 - loss: 0.6563 - accuracy: 0.61 - ETA: 4:29 - loss: 0.6539 - accuracy: 0.62 - ETA: 4:31 - loss: 0.6522 - accuracy: 0.63 - ETA: 4:31 - loss: 0.6486 - accuracy: 0.63 - ETA: 4:31 - loss: 0.6499 - ac

 3720/10000 [==========>...................] - ETA: 4:34 - loss: 0.6756 - accuracy: 0.55 - ETA: 4:38 - loss: 0.6829 - accuracy: 0.55 - ETA: 4:35 - loss: 0.6960 - accuracy: 0.56 - ETA: 4:35 - loss: 0.6668 - accuracy: 0.61 - ETA: 4:33 - loss: 0.6658 - accuracy: 0.62 - ETA: 4:34 - loss: 0.6716 - accuracy: 0.62 - ETA: 4:33 - loss: 0.6703 - accuracy: 0.62 - ETA: 4:31 - loss: 0.6625 - accuracy: 0.62 - ETA: 4:36 - loss: 0.6592 - accuracy: 0.62 - ETA: 4:40 - loss: 0.6532 - accuracy: 0.63 - ETA: 4:43 - loss: 0.6524 - accuracy: 0.62 - ETA: 4:43 - loss: 0.6558 - accuracy: 0.62 - ETA: 4:43 - loss: 0.6588 - accuracy: 0.61 - ETA: 4:41 - loss: 0.6585 - accuracy: 0.61 - ETA: 4:40 - loss: 0.6623 - accuracy: 0.60 - ETA: 4:39 - loss: 0.6585 - accuracy: 0.60 - ETA: 4:37 - loss: 0.6640 - accuracy: 0.60 - ETA: 4:36 - loss: 0.6631 - accuracy: 0.61 - ETA: 4:35 - loss: 0.6585 - accuracy: 0.62 - ETA: 4:34 - loss: 0.6572 - accuracy: 0.61 - ETA: 4:35 - loss: 0.6522 - accuracy: 0.62 - ETA: 4:35 - loss: 0.6515 - ac

10000/10000 [==============================] - ETA: 1:12 - loss: 0.6421 - accuracy: 0.63 - ETA: 1:12 - loss: 0.6421 - accuracy: 0.63 - ETA: 1:11 - loss: 0.6418 - accuracy: 0.63 - ETA: 1:11 - loss: 0.6417 - accuracy: 0.63 - ETA: 1:10 - loss: 0.6417 - accuracy: 0.63 - ETA: 1:10 - loss: 0.6415 - accuracy: 0.63 - ETA: 1:09 - loss: 0.6414 - accuracy: 0.63 - ETA: 1:08 - loss: 0.6415 - accuracy: 0.63 - ETA: 1:08 - loss: 0.6412 - accuracy: 0.63 - ETA: 1:07 - loss: 0.6412 - accuracy: 0.63 - ETA: 1:07 - loss: 0.6411 - accuracy: 0.63 - ETA: 1:06 - loss: 0.6414 - accuracy: 0.63 - ETA: 1:05 - loss: 0.6415 - accuracy: 0.63 - ETA: 1:05 - loss: 0.6415 - accuracy: 0.63 - ETA: 1:04 - loss: 0.6420 - accuracy: 0.63 - ETA: 1:04 - loss: 0.6419 - accuracy: 0.63 - ETA: 1:03 - loss: 0.6420 - accuracy: 0.63 - ETA: 1:03 - loss: 0.6421 - accuracy: 0.63 - ETA: 1:02 - loss: 0.6421 - accuracy: 0.63 - ETA: 1:01 - loss: 0.6420 - accuracy: 0.63 - ETA: 1:01 - loss: 0.6423 - accuracy: 0.63 - ETA: 1:00 - loss: 0.6428 - ac

 3720/10000 [==========>...................] - ETA: 4:48 - loss: 0.6385 - accuracy: 0.60 - ETA: 5:08 - loss: 0.6268 - accuracy: 0.65 - ETA: 5:10 - loss: 0.6321 - accuracy: 0.63 - ETA: 5:08 - loss: 0.6141 - accuracy: 0.65 - ETA: 5:01 - loss: 0.6241 - accuracy: 0.62 - ETA: 4:56 - loss: 0.6479 - accuracy: 0.57 - ETA: 4:53 - loss: 0.6409 - accuracy: 0.58 - ETA: 4:51 - loss: 0.6310 - accuracy: 0.60 - ETA: 4:49 - loss: 0.6295 - accuracy: 0.61 - ETA: 4:48 - loss: 0.6293 - accuracy: 0.62 - ETA: 4:45 - loss: 0.6307 - accuracy: 0.63 - ETA: 4:43 - loss: 0.6317 - accuracy: 0.63 - ETA: 4:43 - loss: 0.6292 - accuracy: 0.63 - ETA: 4:41 - loss: 0.6329 - accuracy: 0.61 - ETA: 4:39 - loss: 0.6415 - accuracy: 0.60 - ETA: 4:38 - loss: 0.6485 - accuracy: 0.59 - ETA: 4:37 - loss: 0.6490 - accuracy: 0.59 - ETA: 4:36 - loss: 0.6550 - accuracy: 0.58 - ETA: 4:35 - loss: 0.6608 - accuracy: 0.57 - ETA: 4:33 - loss: 0.6580 - accuracy: 0.57 - ETA: 4:33 - loss: 0.6584 - accuracy: 0.57 - ETA: 4:32 - loss: 0.6540 - ac

 3720/10000 [==========>...................] - ETA: 4:32 - loss: 0.7686 - accuracy: 0.55 - ETA: 4:33 - loss: 0.6794 - accuracy: 0.65 - ETA: 4:33 - loss: 0.6492 - accuracy: 0.65 - ETA: 4:34 - loss: 0.6290 - accuracy: 0.67 - ETA: 4:33 - loss: 0.6008 - accuracy: 0.71 - ETA: 4:35 - loss: 0.5914 - accuracy: 0.70 - ETA: 4:34 - loss: 0.5878 - accuracy: 0.71 - ETA: 4:35 - loss: 0.5877 - accuracy: 0.70 - ETA: 4:36 - loss: 0.6013 - accuracy: 0.67 - ETA: 4:36 - loss: 0.5960 - accuracy: 0.68 - ETA: 4:35 - loss: 0.5948 - accuracy: 0.68 - ETA: 4:36 - loss: 0.5828 - accuracy: 0.69 - ETA: 4:36 - loss: 0.5873 - accuracy: 0.68 - ETA: 4:37 - loss: 0.5833 - accuracy: 0.69 - ETA: 4:40 - loss: 0.5873 - accuracy: 0.68 - ETA: 4:41 - loss: 0.5914 - accuracy: 0.69 - ETA: 4:43 - loss: 0.6111 - accuracy: 0.67 - ETA: 4:42 - loss: 0.6154 - accuracy: 0.67 - ETA: 4:41 - loss: 0.6179 - accuracy: 0.66 - ETA: 4:40 - loss: 0.6188 - accuracy: 0.66 - ETA: 4:40 - loss: 0.6162 - accuracy: 0.67 - ETA: 4:39 - loss: 0.6157 - ac

10000/10000 [==============================] - ETA: 1:12 - loss: 0.6138 - accuracy: 0.67 - ETA: 1:12 - loss: 0.6141 - accuracy: 0.66 - ETA: 1:11 - loss: 0.6139 - accuracy: 0.67 - ETA: 1:11 - loss: 0.6139 - accuracy: 0.67 - ETA: 1:10 - loss: 0.6139 - accuracy: 0.67 - ETA: 1:10 - loss: 0.6138 - accuracy: 0.67 - ETA: 1:09 - loss: 0.6138 - accuracy: 0.67 - ETA: 1:08 - loss: 0.6136 - accuracy: 0.67 - ETA: 1:08 - loss: 0.6137 - accuracy: 0.67 - ETA: 1:07 - loss: 0.6138 - accuracy: 0.67 - ETA: 1:07 - loss: 0.6137 - accuracy: 0.67 - ETA: 1:06 - loss: 0.6137 - accuracy: 0.67 - ETA: 1:05 - loss: 0.6136 - accuracy: 0.67 - ETA: 1:05 - loss: 0.6139 - accuracy: 0.67 - ETA: 1:04 - loss: 0.6139 - accuracy: 0.67 - ETA: 1:04 - loss: 0.6136 - accuracy: 0.67 - ETA: 1:03 - loss: 0.6137 - accuracy: 0.67 - ETA: 1:03 - loss: 0.6135 - accuracy: 0.67 - ETA: 1:02 - loss: 0.6138 - accuracy: 0.67 - ETA: 1:01 - loss: 0.6135 - accuracy: 0.67 - ETA: 1:01 - loss: 0.6138 - accuracy: 0.67 - ETA: 1:00 - loss: 0.6140 - ac

 3720/10000 [==========>...................] - ETA: 4:31 - loss: 0.5566 - accuracy: 0.70 - ETA: 4:39 - loss: 0.5334 - accuracy: 0.70 - ETA: 4:35 - loss: 0.5531 - accuracy: 0.68 - ETA: 4:36 - loss: 0.5468 - accuracy: 0.68 - ETA: 4:42 - loss: 0.5896 - accuracy: 0.62 - ETA: 4:48 - loss: 0.6213 - accuracy: 0.58 - ETA: 4:54 - loss: 0.6080 - accuracy: 0.62 - ETA: 4:57 - loss: 0.5888 - accuracy: 0.65 - ETA: 4:55 - loss: 0.6013 - accuracy: 0.63 - ETA: 4:52 - loss: 0.5905 - accuracy: 0.65 - ETA: 4:51 - loss: 0.5847 - accuracy: 0.65 - ETA: 4:49 - loss: 0.5965 - accuracy: 0.65 - ETA: 4:47 - loss: 0.5951 - accuracy: 0.66 - ETA: 4:45 - loss: 0.5878 - accuracy: 0.67 - ETA: 4:44 - loss: 0.5872 - accuracy: 0.67 - ETA: 4:42 - loss: 0.5903 - accuracy: 0.67 - ETA: 4:40 - loss: 0.5858 - accuracy: 0.67 - ETA: 4:39 - loss: 0.6126 - accuracy: 0.66 - ETA: 4:38 - loss: 0.6104 - accuracy: 0.66 - ETA: 4:36 - loss: 0.6092 - accuracy: 0.67 - ETA: 4:35 - loss: 0.6034 - accuracy: 0.67 - ETA: 4:34 - loss: 0.6025 - ac

 3720/10000 [==========>...................] - ETA: 4:35 - loss: 0.8424 - accuracy: 0.65 - ETA: 4:34 - loss: 0.6797 - accuracy: 0.75 - ETA: 4:34 - loss: 0.6035 - accuracy: 0.75 - ETA: 4:33 - loss: 0.6084 - accuracy: 0.71 - ETA: 4:32 - loss: 0.6053 - accuracy: 0.74 - ETA: 4:31 - loss: 0.5731 - accuracy: 0.75 - ETA: 4:31 - loss: 0.5902 - accuracy: 0.73 - ETA: 4:30 - loss: 0.5683 - accuracy: 0.75 - ETA: 4:30 - loss: 0.5863 - accuracy: 0.75 - ETA: 4:29 - loss: 0.5774 - accuracy: 0.75 - ETA: 4:29 - loss: 0.5651 - accuracy: 0.75 - ETA: 4:29 - loss: 0.5675 - accuracy: 0.74 - ETA: 4:28 - loss: 0.5646 - accuracy: 0.75 - ETA: 4:27 - loss: 0.5623 - accuracy: 0.75 - ETA: 4:27 - loss: 0.5605 - accuracy: 0.75 - ETA: 4:26 - loss: 0.5587 - accuracy: 0.75 - ETA: 4:26 - loss: 0.5557 - accuracy: 0.75 - ETA: 4:25 - loss: 0.5515 - accuracy: 0.75 - ETA: 4:25 - loss: 0.5508 - accuracy: 0.75 - ETA: 4:24 - loss: 0.5525 - accuracy: 0.75 - ETA: 4:24 - loss: 0.5566 - accuracy: 0.75 - ETA: 4:23 - loss: 0.5587 - ac

10000/10000 [==============================] - ETA: 1:13 - loss: 0.5744 - accuracy: 0.70 - ETA: 1:13 - loss: 0.5744 - accuracy: 0.70 - ETA: 1:12 - loss: 0.5746 - accuracy: 0.70 - ETA: 1:11 - loss: 0.5744 - accuracy: 0.70 - ETA: 1:11 - loss: 0.5740 - accuracy: 0.70 - ETA: 1:10 - loss: 0.5742 - accuracy: 0.70 - ETA: 1:10 - loss: 0.5744 - accuracy: 0.70 - ETA: 1:09 - loss: 0.5749 - accuracy: 0.70 - ETA: 1:09 - loss: 0.5746 - accuracy: 0.70 - ETA: 1:08 - loss: 0.5748 - accuracy: 0.70 - ETA: 1:07 - loss: 0.5746 - accuracy: 0.70 - ETA: 1:07 - loss: 0.5748 - accuracy: 0.70 - ETA: 1:06 - loss: 0.5751 - accuracy: 0.70 - ETA: 1:06 - loss: 0.5750 - accuracy: 0.70 - ETA: 1:05 - loss: 0.5752 - accuracy: 0.70 - ETA: 1:05 - loss: 0.5748 - accuracy: 0.70 - ETA: 1:04 - loss: 0.5745 - accuracy: 0.70 - ETA: 1:03 - loss: 0.5748 - accuracy: 0.70 - ETA: 1:03 - loss: 0.5747 - accuracy: 0.70 - ETA: 1:02 - loss: 0.5745 - accuracy: 0.70 - ETA: 1:02 - loss: 0.5747 - accuracy: 0.70 - ETA: 1:01 - loss: 0.5745 - ac

 3720/10000 [==========>...................] - ETA: 5:18 - loss: 0.8279 - accuracy: 0.60 - ETA: 5:12 - loss: 0.7109 - accuracy: 0.62 - ETA: 5:05 - loss: 0.6716 - accuracy: 0.66 - ETA: 4:56 - loss: 0.6292 - accuracy: 0.70 - ETA: 4:52 - loss: 0.5958 - accuracy: 0.72 - ETA: 4:50 - loss: 0.5944 - accuracy: 0.70 - ETA: 4:47 - loss: 0.6007 - accuracy: 0.68 - ETA: 4:45 - loss: 0.5991 - accuracy: 0.67 - ETA: 4:43 - loss: 0.5886 - accuracy: 0.67 - ETA: 4:43 - loss: 0.5774 - accuracy: 0.69 - ETA: 4:41 - loss: 0.5667 - accuracy: 0.70 - ETA: 4:40 - loss: 0.5738 - accuracy: 0.69 - ETA: 4:40 - loss: 0.5758 - accuracy: 0.69 - ETA: 4:40 - loss: 0.5736 - accuracy: 0.69 - ETA: 4:39 - loss: 0.5817 - accuracy: 0.68 - ETA: 4:38 - loss: 0.5717 - accuracy: 0.69 - ETA: 4:37 - loss: 0.5696 - accuracy: 0.69 - ETA: 4:36 - loss: 0.5760 - accuracy: 0.68 - ETA: 4:36 - loss: 0.5758 - accuracy: 0.68 - ETA: 4:34 - loss: 0.5875 - accuracy: 0.69 - ETA: 4:34 - loss: 0.5841 - accuracy: 0.69 - ETA: 4:33 - loss: 0.5737 - ac

 3720/10000 [==========>...................] - ETA: 4:41 - loss: 0.3218 - accuracy: 0.85 - ETA: 4:39 - loss: 0.3749 - accuracy: 0.85 - ETA: 4:36 - loss: 0.4773 - accuracy: 0.78 - ETA: 4:37 - loss: 0.4751 - accuracy: 0.78 - ETA: 4:35 - loss: 0.4946 - accuracy: 0.77 - ETA: 4:39 - loss: 0.4995 - accuracy: 0.77 - ETA: 4:38 - loss: 0.5090 - accuracy: 0.75 - ETA: 4:37 - loss: 0.4996 - accuracy: 0.77 - ETA: 4:36 - loss: 0.4881 - accuracy: 0.78 - ETA: 4:37 - loss: 0.4939 - accuracy: 0.77 - ETA: 4:36 - loss: 0.5066 - accuracy: 0.75 - ETA: 4:35 - loss: 0.4889 - accuracy: 0.76 - ETA: 4:34 - loss: 0.5003 - accuracy: 0.75 - ETA: 4:33 - loss: 0.5111 - accuracy: 0.74 - ETA: 4:32 - loss: 0.5243 - accuracy: 0.72 - ETA: 4:31 - loss: 0.5309 - accuracy: 0.71 - ETA: 4:30 - loss: 0.5381 - accuracy: 0.70 - ETA: 4:32 - loss: 0.5373 - accuracy: 0.70 - ETA: 4:33 - loss: 0.5350 - accuracy: 0.71 - ETA: 4:34 - loss: 0.5338 - accuracy: 0.71 - ETA: 4:35 - loss: 0.5300 - accuracy: 0.71 - ETA: 4:34 - loss: 0.5308 - ac

10000/10000 [==============================] - ETA: 1:14 - loss: 0.5350 - accuracy: 0.73 - ETA: 1:13 - loss: 0.5348 - accuracy: 0.73 - ETA: 1:13 - loss: 0.5346 - accuracy: 0.73 - ETA: 1:12 - loss: 0.5349 - accuracy: 0.73 - ETA: 1:11 - loss: 0.5346 - accuracy: 0.73 - ETA: 1:11 - loss: 0.5355 - accuracy: 0.73 - ETA: 1:10 - loss: 0.5352 - accuracy: 0.73 - ETA: 1:10 - loss: 0.5353 - accuracy: 0.73 - ETA: 1:09 - loss: 0.5350 - accuracy: 0.73 - ETA: 1:08 - loss: 0.5356 - accuracy: 0.73 - ETA: 1:08 - loss: 0.5360 - accuracy: 0.73 - ETA: 1:07 - loss: 0.5359 - accuracy: 0.73 - ETA: 1:07 - loss: 0.5359 - accuracy: 0.73 - ETA: 1:06 - loss: 0.5360 - accuracy: 0.73 - ETA: 1:05 - loss: 0.5360 - accuracy: 0.73 - ETA: 1:05 - loss: 0.5358 - accuracy: 0.73 - ETA: 1:04 - loss: 0.5357 - accuracy: 0.73 - ETA: 1:04 - loss: 0.5357 - accuracy: 0.73 - ETA: 1:03 - loss: 0.5359 - accuracy: 0.73 - ETA: 1:03 - loss: 0.5361 - accuracy: 0.73 - ETA: 1:02 - loss: 0.5359 - accuracy: 0.73 - ETA: 1:01 - loss: 0.5356 - ac

 3720/10000 [==========>...................] - ETA: 4:46 - loss: 0.6730 - accuracy: 0.60 - ETA: 4:36 - loss: 0.5580 - accuracy: 0.72 - ETA: 4:37 - loss: 0.5698 - accuracy: 0.71 - ETA: 4:43 - loss: 0.5376 - accuracy: 0.75 - ETA: 4:50 - loss: 0.5500 - accuracy: 0.74 - ETA: 4:52 - loss: 0.5357 - accuracy: 0.75 - ETA: 4:54 - loss: 0.5024 - accuracy: 0.77 - ETA: 4:51 - loss: 0.4991 - accuracy: 0.77 - ETA: 4:54 - loss: 0.5060 - accuracy: 0.76 - ETA: 4:57 - loss: 0.5186 - accuracy: 0.76 - ETA: 4:55 - loss: 0.5288 - accuracy: 0.75 - ETA: 4:54 - loss: 0.5170 - accuracy: 0.76 - ETA: 4:52 - loss: 0.5151 - accuracy: 0.76 - ETA: 4:52 - loss: 0.5092 - accuracy: 0.76 - ETA: 4:51 - loss: 0.4995 - accuracy: 0.76 - ETA: 4:50 - loss: 0.5017 - accuracy: 0.75 - ETA: 4:49 - loss: 0.4892 - accuracy: 0.76 - ETA: 4:50 - loss: 0.4863 - accuracy: 0.77 - ETA: 4:50 - loss: 0.4896 - accuracy: 0.77 - ETA: 4:50 - loss: 0.4925 - accuracy: 0.77 - ETA: 4:50 - loss: 0.4916 - accuracy: 0.77 - ETA: 4:50 - loss: 0.4891 - ac

 3720/10000 [==========>...................] - ETA: 4:17 - loss: 0.4409 - accuracy: 0.80 - ETA: 4:16 - loss: 0.4914 - accuracy: 0.77 - ETA: 4:27 - loss: 0.4661 - accuracy: 0.78 - ETA: 4:28 - loss: 0.5467 - accuracy: 0.75 - ETA: 4:27 - loss: 0.4994 - accuracy: 0.78 - ETA: 4:28 - loss: 0.5058 - accuracy: 0.77 - ETA: 4:28 - loss: 0.5045 - accuracy: 0.77 - ETA: 4:29 - loss: 0.4868 - accuracy: 0.78 - ETA: 4:28 - loss: 0.4810 - accuracy: 0.77 - ETA: 4:28 - loss: 0.4799 - accuracy: 0.78 - ETA: 4:27 - loss: 0.4857 - accuracy: 0.77 - ETA: 4:27 - loss: 0.4812 - accuracy: 0.78 - ETA: 4:27 - loss: 0.4723 - accuracy: 0.79 - ETA: 4:27 - loss: 0.4796 - accuracy: 0.78 - ETA: 4:26 - loss: 0.4684 - accuracy: 0.79 - ETA: 4:26 - loss: 0.4650 - accuracy: 0.80 - ETA: 4:25 - loss: 0.4675 - accuracy: 0.79 - ETA: 4:25 - loss: 0.4704 - accuracy: 0.79 - ETA: 4:24 - loss: 0.4601 - accuracy: 0.80 - ETA: 4:24 - loss: 0.4625 - accuracy: 0.79 - ETA: 4:23 - loss: 0.4774 - accuracy: 0.79 - ETA: 4:23 - loss: 0.4843 - ac

10000/10000 [==============================] - ETA: 1:12 - loss: 0.4779 - accuracy: 0.77 - ETA: 1:12 - loss: 0.4781 - accuracy: 0.77 - ETA: 1:11 - loss: 0.4775 - accuracy: 0.77 - ETA: 1:11 - loss: 0.4773 - accuracy: 0.77 - ETA: 1:10 - loss: 0.4767 - accuracy: 0.77 - ETA: 1:10 - loss: 0.4764 - accuracy: 0.77 - ETA: 1:09 - loss: 0.4765 - accuracy: 0.77 - ETA: 1:08 - loss: 0.4780 - accuracy: 0.77 - ETA: 1:08 - loss: 0.4780 - accuracy: 0.77 - ETA: 1:07 - loss: 0.4779 - accuracy: 0.77 - ETA: 1:07 - loss: 0.4776 - accuracy: 0.77 - ETA: 1:06 - loss: 0.4777 - accuracy: 0.77 - ETA: 1:05 - loss: 0.4778 - accuracy: 0.77 - ETA: 1:05 - loss: 0.4778 - accuracy: 0.77 - ETA: 1:04 - loss: 0.4779 - accuracy: 0.77 - ETA: 1:04 - loss: 0.4776 - accuracy: 0.77 - ETA: 1:03 - loss: 0.4777 - accuracy: 0.77 - ETA: 1:03 - loss: 0.4776 - accuracy: 0.77 - ETA: 1:02 - loss: 0.4775 - accuracy: 0.77 - ETA: 1:01 - loss: 0.4772 - accuracy: 0.77 - ETA: 1:01 - loss: 0.4774 - accuracy: 0.77 - ETA: 1:00 - loss: 0.4778 - ac

 3720/10000 [==========>...................] - ETA: 5:36 - loss: 0.4451 - accuracy: 0.80 - ETA: 5:36 - loss: 0.5255 - accuracy: 0.77 - ETA: 6:02 - loss: 0.4871 - accuracy: 0.81 - ETA: 6:00 - loss: 0.4590 - accuracy: 0.81 - ETA: 5:58 - loss: 0.4956 - accuracy: 0.78 - ETA: 5:52 - loss: 0.5025 - accuracy: 0.77 - ETA: 5:45 - loss: 0.4987 - accuracy: 0.76 - ETA: 5:43 - loss: 0.4820 - accuracy: 0.77 - ETA: 5:42 - loss: 0.5136 - accuracy: 0.75 - ETA: 5:40 - loss: 0.5111 - accuracy: 0.74 - ETA: 5:35 - loss: 0.5037 - accuracy: 0.75 - ETA: 5:32 - loss: 0.5104 - accuracy: 0.74 - ETA: 5:28 - loss: 0.5080 - accuracy: 0.74 - ETA: 5:30 - loss: 0.5052 - accuracy: 0.74 - ETA: 5:28 - loss: 0.5063 - accuracy: 0.74 - ETA: 5:26 - loss: 0.4959 - accuracy: 0.75 - ETA: 5:24 - loss: 0.4901 - accuracy: 0.75 - ETA: 5:20 - loss: 0.4876 - accuracy: 0.75 - ETA: 5:17 - loss: 0.4820 - accuracy: 0.75 - ETA: 5:14 - loss: 0.4822 - accuracy: 0.75 - ETA: 5:11 - loss: 0.4848 - accuracy: 0.75 - ETA: 5:10 - loss: 0.4882 - ac

 3720/10000 [==========>...................] - ETA: 4:21 - loss: 0.3189 - accuracy: 0.85 - ETA: 4:26 - loss: 0.3408 - accuracy: 0.87 - ETA: 4:28 - loss: 0.3841 - accuracy: 0.83 - ETA: 4:31 - loss: 0.4704 - accuracy: 0.78 - ETA: 4:30 - loss: 0.4407 - accuracy: 0.80 - ETA: 4:29 - loss: 0.4520 - accuracy: 0.80 - ETA: 4:30 - loss: 0.4294 - accuracy: 0.82 - ETA: 4:29 - loss: 0.4572 - accuracy: 0.79 - ETA: 4:29 - loss: 0.4471 - accuracy: 0.78 - ETA: 4:29 - loss: 0.4496 - accuracy: 0.79 - ETA: 4:31 - loss: 0.4415 - accuracy: 0.79 - ETA: 4:35 - loss: 0.4547 - accuracy: 0.80 - ETA: 4:37 - loss: 0.4719 - accuracy: 0.78 - ETA: 4:38 - loss: 0.4660 - accuracy: 0.78 - ETA: 4:37 - loss: 0.4608 - accuracy: 0.79 - ETA: 4:36 - loss: 0.4587 - accuracy: 0.78 - ETA: 4:35 - loss: 0.4609 - accuracy: 0.78 - ETA: 4:34 - loss: 0.4649 - accuracy: 0.78 - ETA: 4:33 - loss: 0.4717 - accuracy: 0.77 - ETA: 4:32 - loss: 0.4700 - accuracy: 0.78 - ETA: 4:31 - loss: 0.4697 - accuracy: 0.78 - ETA: 4:30 - loss: 0.4726 - ac

10000/10000 [==============================] - ETA: 1:15 - loss: 0.4457 - accuracy: 0.79 - ETA: 1:14 - loss: 0.4454 - accuracy: 0.79 - ETA: 1:13 - loss: 0.4459 - accuracy: 0.79 - ETA: 1:13 - loss: 0.4458 - accuracy: 0.79 - ETA: 1:12 - loss: 0.4464 - accuracy: 0.79 - ETA: 1:12 - loss: 0.4459 - accuracy: 0.79 - ETA: 1:11 - loss: 0.4455 - accuracy: 0.79 - ETA: 1:10 - loss: 0.4451 - accuracy: 0.79 - ETA: 1:10 - loss: 0.4449 - accuracy: 0.79 - ETA: 1:09 - loss: 0.4448 - accuracy: 0.79 - ETA: 1:09 - loss: 0.4449 - accuracy: 0.79 - ETA: 1:08 - loss: 0.4457 - accuracy: 0.79 - ETA: 1:07 - loss: 0.4452 - accuracy: 0.79 - ETA: 1:07 - loss: 0.4448 - accuracy: 0.79 - ETA: 1:06 - loss: 0.4447 - accuracy: 0.79 - ETA: 1:06 - loss: 0.4449 - accuracy: 0.79 - ETA: 1:05 - loss: 0.4446 - accuracy: 0.79 - ETA: 1:05 - loss: 0.4443 - accuracy: 0.79 - ETA: 1:04 - loss: 0.4438 - accuracy: 0.79 - ETA: 1:04 - loss: 0.4438 - accuracy: 0.79 - ETA: 1:03 - loss: 0.4441 - accuracy: 0.79 - ETA: 1:02 - loss: 0.4445 - ac

### Load the Model with the Best Validation Loss

In [14]:
model.load_weights('input/saved_models/weights.best.from_scratch.hdf5')

### Test the Model

Trying out the model on the test dataset images.

In [15]:
# get index of predicted smile detection for each image in test set
smile_predictions = [np.argmax(model.predict(np.expand_dims(tensor, axis=0))) for tensor in test_tensors]

# report test accuracy
test_accuracy = 100*np.sum(np.array(smile_predictions)==np.argmax(test_targets, axis=1))/len(smile_predictions)
print('Test accuracy: %.4f%%' % test_accuracy)

Test accuracy: 81.0400%



Here, we see that a simple CNN model is developed and that its accuracy is 81.04%. This is considered as a Benchmark and we try to develop a model with a better accuracy considering augmentation, transfer learning, bottlenexk features, pretuning, etc in the other notebooks.